packages/web-frontend/cypress/scripts/testE2e/Snapshots.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import fs from 'fs';
+import { isPlainObject, set } from 'lodash';
+import { extname } from 'path';
+
+import { compareAsc } from '@tupaia/utils';
+
+/**
+ * NodePath is an array of two elements:
+ * NodePath[0]: array of paths to reach a certain node in a tree
+ * NodePath[1]: value of the node
+ *
+ * @typedef {[string[], any]} NodePath
+ */
+
+const nodePathsToTree = nodePaths =>
+  nodePaths.reduce((agg, entry) => {
+    const [path, value] = entry;
+    set(agg, path, value);
+    return agg;
+  }, {});
+
+const isJs = path => extname(path) === '.js';
+
+export class Snapshots {
+  snapshotTree;
+
+  /**
+   * Method-specific, used during snapshot extraction
+   */
+  extractionConfig = {};
+
+  /**
+   * @param {Object<string, any>} snapshotTree
+   */
+  constructor(snapshotTree) {
+    this.snapshotTree = snapshotTree;
+  }
+
+  static import = path => {
+    const contents = fs.readFileSync(path, { encoding: 'utf-8' });
+    // Note: JS parsing is a bit naive here, ideally we should use a code parsing library
+    const snapshotJson = isJs(path) ? contents.replace('module.exports = ', '') : contents;
+    const snapshotTree = JSON.parse(snapshotJson);
+    return new Snapshots(snapshotTree);
+  };
+
+  export = path => {
+    const contents = isJs(path) ? `module.exports = ${this.toString()}` : this.toString();
+    fs.writeFileSync(path, contents);
+  };
+
+  toString = () => JSON.stringify(this.snapshotTree, undefined, 2);
+
+  equals = snapshots => {
+    if (!(snapshots instanceof Snapshots)) {
+      throw new Error('"Snapshots.equals" expects a Snapshots instance');
+    }
+    return this.toString() === snapshots.toString();
+  };
+
+  isEmpty = () => Object.keys(this.snapshotTree).length === 0;
+
+  extractSnapshotsByKey = (key, options = {}) => {
+    const extractedNodePaths = [];
+    this.extractionConfig = {
```

Comment:
Can we pass through `extractionConfig` as a third argument, rather than storing it on the instance?

Reply:
We definitely can! It's a matter of preference. I thought it would be redundant to just pass the same value over and over in all recursions.

I'm personally in the "a little state doesn't harm" school of thought, although I guess it's a bit unconventional these days. Very happy to change that here if you prefer!

Reply:
A bit more explanation on my reasoning: when I say "redundant", I mean in relation to code simplicity (in function signatures), not to performance toll (which would negligible, if any). The tradeoff is: more complexity in the class state level for less complexity in the recursion logic (signature). Recursions are generally more complicated than average, so I thought the tradeoff is worth it.

Reply:
I find it unexpected, so confusing because I have to put the pieces together. I agree that recursion is confusing, but I think this nice static config parameter shouldn't add much mental load, whereas understanding the method-specific state did for me. It may be a personal preference thing though, so will leave it up to you!

Reply:
No worries, I'll change! Your suggestion probably adheres to the principle of least surprise in the context of our codebase - and is certainly clean!

That was a bit of an experimental approach so I mainly wanted to understand your viewpoint 😉